ops/testing.py

```diff
@@ -2586,6 +2605,128 @@ class NonAbsolutePathError(Exception):
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def push_path(self,
+                  source_path: Union[StrOrPath, Iterable[StrOrPath]],
+                  dest_dir: StrOrPath) -> None:
+        """Recursively push a local path or files to the remote system.
+
+        See :meth:`ops.model.Container.push_path` for details.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def pull_path(self,
+                  source_path: Union[StrOrPath, Iterable[StrOrPath]],
+                  dest_dir: StrOrPath) -> None:
+        """Recursively pull a remote path or files to the local system.
+
+        See :meth:`ops.model.Container.pull_path` for details.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def exists(self, path: str) -> bool:
+        """Return true if the path exists on the container filesystem.
+
+        See :meth:`ops.model.Container.exists` for details.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def isdir(self, path: str) -> bool:
+        """Return true if a directory exists at the given path on the container filesystem.
+
+        See :meth:`ops.model.Container.isdir` for details.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def make_dir(
+            self, path: str, *, make_parents: bool = False, permissions: Optional[int] = None,
+            user_id: Optional[int] = None, user: Optional[str] = None,
+            group_id: Optional[int] = None, group: Optional[str] = None) -> None:
+        """Create a directory on the remote system with the given attributes.
+
+        See :meth:`ops.model.Container.make_dir` for details.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def remove_path(self, path: str, *, recursive: bool = False) -> None:
+        """Remove a file or directory on the remote system.
+
+        See :meth:`ops.model.Container.remove_path` for details.
+        """
+        raise NotImplementedError
+
+
+class _ContainerFilesystemProxy:
+    """A proxy providing unrestricted filesystem access for a specified container."""
+
+    def __init__(self, container: model.Container, backend: _TestingPebbleClient):
+        self._container = container
+        self._backend = backend
+
+    def __getattr__(self, name: str):
+        if name not in ContainerFilesystem.__dict__ or name.startswith("_"):
+            raise AttributeError(f'{self.__class__.__name__!r} object has no attribute {name!r}')
+        func = getattr(self._container, name)
+
+        @functools.wraps(func)
+        def wrapper(*args: Any, **kwargs: Any) -> Any:
+            can_connect = self._backend._can_connect(self._container._pebble)
+            self._backend._set_can_connect(self._container._pebble, True)
+            try:
+                return func(*args, **kwargs)
+            finally:
+                self._backend._set_can_connect(self._container._pebble, can_connect)
```

Comment:
Well, I can certainly achieve this without mutating the model state, by temporary hijacking the `_check_connection` in the test pebble client and disabling all the checks for the filesystem access.

```python
        @functools.wraps(func)
        def wrapper(*args: Any, **kwargs: Any) -> Any:
            pebble_client = self._container._pebble
            with unittest.mock.patch.object(target=pebble_client, attribute="_check_connection"):
                return func(*args, **kwargs)
```

Do you think this could make it better?

Reply:
The incentive to mimic the API of the container for the API of the container filesystem is that users will not need to learn a new API just for testing; they can reuse their knowledge of the container API.